File: multicore-f/L2/solver/src/mc.f
!C    
!C***
!C*** MC
!C***
!C
!C    Multicolor Ordering Method
!C    

      subroutine MC (N, NL, NU, INL, IAL, INU, IAU,                     &
     &               NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW)

      implicit REAL*8(A-H,O-Z)

      integer, dimension(N)   :: INL, INU, NEWtoOLD, OLDtoNEW
      integer, dimension(0:N) :: COLORindex
      integer, dimension(NL,N):: IAL
      integer, dimension(NU,N):: IAU

      integer, dimension(:)  , allocatable :: IW, INLw, INUw
      integer, dimension(:,:), allocatable :: IALw, IAUw

!C
!C +-------+
!C | INIT. |
!C +-------+
!C===
      allocate (IW(N))
      IW= 0

      NCOLORk  = NCOLORtot

      do i= 1, N
        NEWtoOLD  (i)= i
        OLDtoNEW  (i)= i
      enddo

       INmin= N
      NODmin= 0
      do i= 1, N
        icon= INL(i) + INU(i)
        if (icon.lt.INmin) then
           INmin= icon
          NODmin= i
        endif
      enddo

      OLDtoNEW(NODmin)= 1
      NEWtoOLD(     1)= NODmin

      IW        = 0
      IW(NODmin)= 1

      ITEMcou= N/NCOLORk
!C===

!C
!C +---------------+
!C | MULTIcoloring |
!C +---------------+
!C===
      icou = 1
      icouK= 1

      do icol= 1, N
        NCOLORk= icol
        do i= 1, N
          if (IW(i).le.0) IW(i)= 0
        enddo
        do i= 1, N
!C
!C-- already COLORED
          if (IW(i).eq.icol) then
            do k= 1, INL(i)
              ik= IAL(k,i)
              if (IW(ik).le.0) IW(ik)= -1
            enddo
            do k= 1, INU(i)
              ik= IAU(k,i)
              if (IW(ik).le.0) IW(ik)= -1
            enddo
          endif
!C
!C-- not COLORED
          if (IW(i).eq.0) then
            icou = icou  + 1
            icouK= icouK + 1
            IW(i)= icol
            do k= 1, INL(i)
              ik= IAL(k,i)
              if (IW(ik).le.0) IW(ik)= -1
            enddo
            do k= 1, INU(i)
              ik= IAU(k,i)
              if (IW(ik).le.0) IW(ik)= -1
            enddo
          endif
          if (icou .eq.N)       goto 200
          if (icouK.eq.ITEMcou) goto 100
        enddo      
 100    continue
        icouK= 0
      enddo

 200  continue
!C===

!C
!C +----------------+
!C | FINAL COLORING |
!C +----------------+
!C===
      NCOLORtot= NCOLORk
      COLORindex= 0
      icoug= 0
      do ic= 1, NCOLORtot
        icou= 0
        do i= 1, N
          if (IW(i).eq.ic) then
            icou = icou + 1
            icoug= icoug + 1
            NEWtoOLD(icoug)= i
            OLDtoNEW(i    )= icoug
          endif
        enddo
        COLORindex(ic)= icou
      enddo

      do ic= 1, NCOLORtot
        COLORindex(ic)= COLORindex(ic-1) + COLORindex(ic)
      enddo
!C===

!C
!C +-----------------+
!C | MATRIX transfer |
!C +-----------------+
!C===
      allocate (INLw(N), INUw(N), IALw(NL,N), IAUw(NU,N))
      do j= 1, NL
        do i= 1, N
          IW(i) = IAL(j,NEWtoOLD(i))
        enddo
        do i= 1, N
          IAL(j,i) = IW(i)
        enddo
      enddo

      do j= 1, NU
        do i= 1, N
          IW(i) = IAU(j,NEWtoOLD(i))
        enddo
        do i= 1, N
          IAU(j,i) = IW(i)
        enddo
      enddo

      do i= 1, N
        IW(i) = INL(NEWtoOLD(i))
      enddo

      do i= 1, N
        INLw(i) = IW(i)
      enddo

      do i= 1, N
        IW(i) = INU(NEWtoOLD(i))
      enddo

      do i= 1, N
        INUw(i) = IW(i)
      enddo

      do j= 1, NL
        do i= 1, N
          if (IAL(j,i).eq.0) then
            IALw(j,i) = 0
           else
            IALw(j,i) = OLDtoNEW(IAL(j,i))
          endif
        enddo
      enddo

      do j= 1, NU
        do i= 1, N
          if ( IAU(j,i).eq.0) then
            IAUw(j,i) = 0
           else
            IAUw(j,i) = OLDtoNEW(IAU(j,i))
          endif
        enddo
      enddo

      INL= 0
      INU= 0
      IAL= 0
      IAU= 0

      do i= 1, N
        jL= 0
        jU= 0
        do j= 1, INLw(i)
          if (IALw(j,i).gt.i) then
            jU= jU + 1
            IAU(jU,i)= IALw(j,i)
           else
            jL= jL + 1
            IAL(jL,i)= IALw(j,i)
          endif
         enddo

        do j= 1, INUw(i)
          if (IAUw(j,i).gt.i) then
            jU= jU + 1
            IAU(jU,i)= IAUw(j,i)
           else
            jL= jL + 1
            IAL(jL,i)= IAUw(j,i)
          endif
         enddo

         INL(i)= jL
         INU(i)= jU
      enddo
!C===
      deallocate (IW, INLw, INUw, IALw, IAUw)

      return
      end

File: multicore-c/L2/coloring/src/mc.c
/*
 * Multicolor Ordering Method
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include "mc.h"

int
MC(int N, int NL, int NU, int *INL, int **IAL, int *INU, int **IAU,
	int *NCOLORtot, int *COLORindex, int *NEWtoOLD, int *OLDtoNEW)
{
	int NCOLORk;
	int *IW, *INLw, *INUw;
	int **IALw, **IAUw;
	int INmin, NODmin, ITEMcou;
	int i, j, k, icon, icou, icouK, icoug, icol, ic, ik, jL, jU;

/*********
 * INIT. *
 *********/
	IW = (int *)calloc(N, sizeof(int));
	if(IW == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}

	NCOLORk = *NCOLORtot;

	for(i=0; i<N; i++) {
		NEWtoOLD[i] = i+1;
		OLDtoNEW[i] = i+1;
	}

	INmin = N;
	NODmin = -1;
	for(i=0; i<N; i++) {
		icon = INL[i] + INU[i];
		if (icon < INmin) {
			INmin = icon;
			NODmin = i;
		}
	}

	OLDtoNEW[NODmin] = 1;
	NEWtoOLD[0     ] = 1+NODmin;

	memset(IW, 0, sizeof(int)*N);
	IW[NODmin] = 1;

	ITEMcou = N / NCOLORk;

/*****************
 * MULTIcoloring *
 *****************/
	icou = 1;
	icouK = 1;
	for(icol=0; icol<N; icol++) {
		NCOLORk = icol + 1;
		for(i=0; i<N; i++) {
			if(IW[i] <= 0) IW[i] = 0;
		}
		for(i=0; i<N; i++) {
/*
 * already COLORED
 */
			if(IW[i] == icol+1) {
				for(k=0; k<INL[i]; k++) {
					ik = IAL[i][k];
					if(IW[ik-1] <= 0) IW[ik-1] = -1;
				}
				for(k=0; k<INU[i]; k++) {
					ik = IAU[i][k];
					if(IW[ik-1] <= 0) IW[ik-1] = -1;
				}
			}

/*
 * not COLORED
 */
			if(IW[i] == 0) {
				icou++;
				icouK++;

				IW[i] = icol + 1;
				for(k=0; k<INL[i]; k++) {
					ik = IAL[i][k];
					if(IW[ik-1] <= 0) IW[ik-1] = -1;
				}
				for(k=0; k<INU[i]; k++) {
					ik = IAU[i][k];
					if(IW[ik-1] <= 0) IW[ik-1] = -1;
				}
			}
			if(icou == N) goto N200;
			if(icouK == ITEMcou) goto N100;
		}
N100:
		icouK = 0;
	}
N200:

/******************
 * Final Coloring *
 ******************/
	*NCOLORtot = NCOLORk;
	memset(COLORindex, 0, sizeof(int)*(N+1));
	icoug = 0;
	for(ic=1; ic<=(*NCOLORtot); ic++) {
		icou = 0;
		for(i=0; i<N; i++) {
			if(IW[i] == ic) {
				NEWtoOLD[icoug] = 1+i;
				OLDtoNEW[i]     = 1+icoug;
				icou++;
				icoug++;
			}
		}
		COLORindex[ic] = icou;
	}

	for(ic=1; ic<=(*NCOLORtot); ic++) {
		COLORindex[ic] += COLORindex[ic-1];
	}

/*******************
 * MATRIX transfer *
 *******************/
	INLw = (int *)calloc(N, sizeof(int));
	if(INLw == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	INUw = (int *)calloc(N, sizeof(int));
	if(INUw == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	IALw = (int **)calloc(N, sizeof(int *));
	if(IALw == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	for(i=0; i<N; i++) {
		IALw[i] = (int *)calloc(NL, sizeof(int));
		if(IALw[i] == NULL) {
			fprintf(stderr, "Error: %s\n", strerror(errno));
			return -1;
		}
	}
	IAUw = (int **)calloc(N, sizeof(int *));
	if(IAUw == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	for(i=0; i<N; i++) {
		IAUw[i] = (int *)calloc(NU, sizeof(int *));
		if(IAUw[i] == NULL) {
			fprintf(stderr, "Error: %s\n", strerror(errno));
			return -1;
		}
	}

	for(j=0; j<NL; j++) {
		for(i=0; i<N; i++) {
			IW[i] = IAL[NEWtoOLD[i]-1][j];
		}
		for(i=0; i<N; i++) {
			IAL[i][j] = IW[i];
		}
	}

	for(j=0; j<NU; j++) {
		for(i=0; i<N; i++) {
			IW[i] = IAU[NEWtoOLD[i]-1][j];
		}
		for(i=0; i<N; i++) {
			IAU[i][j] = IW[i];
		}
	}

	for(i=0; i<N; i++) {
		IW[i] = INL[NEWtoOLD[i]-1];
	}

	for(i=0; i<N; i++) {
		INLw[i] = IW[i];
	}

	for(i=0; i<N; i++) {
		IW[i] = INU[NEWtoOLD[i]-1];
	}

	for(i=0; i<N; i++) {
		INUw[i] = IW[i];
	}

	for(j=0; j<NL; j++) {
		for(i=0; i<N; i++) {
			if(IAL[i][j] == 0) {
				IALw[i][j] = 0;
			} else {
			  IALw[i][j] = OLDtoNEW[IAL[i][j]-1];
			}
		}
	}

	for(j=0; j<NU; j++) {
		for(i=0; i<N; i++) {
			if(IAU[i][j] == 0) {
				IAUw[i][j] = 0;
			} else {
				IAUw[i][j] = OLDtoNEW[IAU[i][j]-1];
			}
		}
	}

	memset(INL, 0, sizeof(int)*N);
	memset(INU, 0, sizeof(int)*N);
	for(i=0; i<N; i++) {
		memset(IAL[i], 0, sizeof(int)*NL);
	}
	for(i=0; i<N; i++) {
		memset(IAU[i], 0, sizeof(int)*NU);
	}

	for(i=0; i<N; i++) {
		jL = 0;
		jU = 0;

		for(j=0; j<INLw[i]; j++) {
			if(IALw[i][j] > i+1) {
				IAU[i][jU] = IALw[i][j];
				jU++;
			} else {
				IAL[i][jL] = IALw[i][j];
				jL++;
			}
		}

		for(j=0; j<INUw[i]; j++) {
			if(IAUw[i][j] > i+1) {
				IAU[i][jU] = IAUw[i][j];
				jU++;
			} else {
				IAL[i][jL] = IAUw[i][j];
				jL++;
			}
		}

		INL[i] = jL;
		INU[i] = jU;
	}

	free(IW);
	free(INLw);
	free(INUw);
	free(IALw);
	free(IAUw);

	return 0;
}
